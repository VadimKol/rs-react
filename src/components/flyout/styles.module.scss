.flyout {
  position: fixed;
  bottom: 30px;
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 500px;
  padding: 15px;
  background:
    linear-gradient(var(--background-color) 0 0) padding-box,
    linear-gradient(to top, var(--primary-color), var(--secondary-color)) border-box;
  border: 5px solid transparent;
  border-radius: 10px;


  @media screen and (width <= 874px) {
    width: 280px;
  }
}

.title {
  margin-top: 0;
  color: var(--text-color);
}

.buttons {
  display: flex;
  gap: 20px;
}

.button {
  font-size: 1.2rem;

  @media screen and (width <= 874px) {
    font-size: 0.9rem;
  }
}

.download_link {
  text-decoration: none;
  color: var(--text-color);
}

