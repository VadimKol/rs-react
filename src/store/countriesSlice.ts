import { createSlice, type PayloadAction } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';

import { countries } from '@/common/countries';

import { type RootState } from './store';

export const countriesLice = createSlice({
  name: 'countries',
  initialState: countries,
  reducers: {
    addCountry: (state, action: PayloadAction<string>) => {
      state.push(action.payload);
    },
  },
});

export const { addCountry } = countriesLice.actions;

export const selectCountries = (state: RootState): string[] => state.countries;

export const useCountries = (): string[] => useSelector(selectCountries);
