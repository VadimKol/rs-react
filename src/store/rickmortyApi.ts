import { createApi, fakeBaseQuery } from '@reduxjs/toolkit/query/react';
import { getCharacter, getCharacters } from 'rickmortyapi';

export const rickmortyApi = createApi({
  reducerPath: 'rickmortyApi',
  baseQuery: fakeBaseQuery(),
  endpoints: (builder) => ({
    getCharacters: builder.query({
      queryFn: async ({ page, character }: { page: number; character: { name: string } }) => {
        try {
          const { data, status, statusMessage } = await getCharacters({ page, name: character.name });

          if (status === 200) {
            return { data: { characters: data.results || [], totalPages: data.info?.pages || 0 } };
          }
          if (status === 404) {
            return { data: { characters: [], totalPages: 0 } };
          }

          throw new Error(statusMessage);
        } catch (error) {
          return { error };
        }
      },
    }),
    getCharacter: builder.query({
      queryFn: async (id: number) => {
        try {
          const { data, status, statusMessage } = await getCharacter(id);

          if (status === 200) {
            return { data };
          }
          if (status === 404) {
            return { data: null };
          }

          throw new Error(statusMessage);
        } catch (error) {
          return { error };
        }
      },
    }),
  }),
});

export const { useGetCharactersQuery, useGetCharacterQuery } = rickmortyApi;
