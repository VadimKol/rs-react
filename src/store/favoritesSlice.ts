/* import { createSlice, type PayloadAction } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';
import { type Character } from 'rickmortyapi';

import { type RootState } from './store';

const initialState: Character[] = [];

export const favoritesSLice = createSlice({
  name: 'favorites',
  initialState,
  reducers: {
    addToFavorites: (state, action: PayloadAction<Character>) => {
      state.push(action.payload);
    },
    removeFromFavorites: (state, action: PayloadAction<Character>) =>
      state.filter((character) => character.id !== action.payload.id),
    clearFavorites: () => [],
  },
});

export const { addToFavorites, removeFromFavorites, clearFavorites } = favoritesSLice.actions;

const selectFavorites = (state: RootState): Character[] => state.favorites;

const isFavorite = (state: RootState, character?: Character): boolean =>
  state.favorites.some((favorite) => favorite.id === character?.id);

export const useFavorites = (): Character[] => useSelector(selectFavorites);

export const useIsFavorite = (character?: Character): boolean =>
  useSelector((state: RootState) => isFavorite(state, character));
 */
