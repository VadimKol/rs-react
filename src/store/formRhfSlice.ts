import { createSlice, type PayloadAction } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';

import { type FormInfo } from '@/common/types';

import { type RootState } from './store';

const initialState: FormInfo = {
  name: null,
  age: null,
  email: null,
  password: null,
  confirmPassword: null,
  gender: null,
  country: null,
  tc: null,
  image: null,
  strength: '0',
};

export const formRhfSLice = createSlice({
  name: 'formRhf',
  initialState,
  reducers: {
    updateFormRhf: (_, action: PayloadAction<FormInfo>) => action.payload,
  },
});

export const { updateFormRhf } = formRhfSLice.actions;

const selectFormRhfInfo = (state: RootState): FormInfo => state.formRhf;

export const useFormRhfInfo = (): FormInfo => useSelector(selectFormRhfInfo);
