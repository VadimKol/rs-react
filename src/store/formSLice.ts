import { createSlice, type PayloadAction } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';

import { type FormInfo } from '@/common/types';

import { type RootState } from './store';

export const initialState: FormInfo[] = [];

export const formSLice = createSlice({
  name: 'form',
  initialState,
  reducers: {
    updateForm: (state, action: PayloadAction<FormInfo>) => {
      state.push(action.payload);
    },
  },
});

export const { updateForm } = formSLice.actions;

const selectFormInfo = (state: RootState): FormInfo[] => state.form;

export const useFormInfo = (): FormInfo[] => useSelector(selectFormInfo);
