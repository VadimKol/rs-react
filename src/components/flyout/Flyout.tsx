import { type ReactNode, useEffect } from 'react';
import { useDispatch } from 'react-redux';

import { charactersToCsv } from '@/common/utils';
import { clearFavorites, useFavorites } from '@/store/favoritesSlice';

import { CustomButton } from '../custom-button/Ð¡ustomButton';
import styles from './styles.module.scss';

export function Flyout(): ReactNode {
  const dispatch = useDispatch();
  const favorites = useFavorites();
  const url = URL.createObjectURL(new Blob([charactersToCsv(favorites)], { type: 'text/csv' }));

  useEffect(() => (): void => URL.revokeObjectURL(url), [url]);

  return (
    !!favorites.length && (
      <div className={styles.flyout}>
        <h3 className={styles.title}>{`${favorites.length} items are selected`}</h3>
        <div className={styles.buttons}>
          <CustomButton className={styles.button} aria-label="Unselect all" onClick={() => dispatch(clearFavorites())}>
            Unselect all
          </CustomButton>
          <CustomButton className={styles.button} aria-label="Download as .csv">
            <a className={styles.download_link} href={url} download={`${favorites.length}_characters.csv`}>
              Download
            </a>
          </CustomButton>
        </div>
      </div>
    )
  );
}
