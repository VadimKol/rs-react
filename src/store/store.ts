import { configureStore } from '@reduxjs/toolkit';

import { formSLice } from './formSLice';

export const store = configureStore({ reducer: { form: formSLice.reducer } });

export type RootState = ReturnType<typeof store.getState>;
