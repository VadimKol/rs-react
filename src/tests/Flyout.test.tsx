import { render } from '@testing-library/react';
import { Provider } from 'react-redux';

import { Flyout } from '@/components/flyout/Flyout';
import { store } from '@/store/store';

import { characters } from './mocks/data';

global.URL.createObjectURL = vi.fn();
global.URL.revokeObjectURL = vi.fn();

describe('Flyout component', () => {
  it('renders correctly', () => {
    vi.mock('@/store/favoritesSlice', async (importOriginal) => {
      const actual = await importOriginal<object>();
      return {
        ...actual,
        useFavorites: vi.fn(() => characters),
      };
    });

    const { container } = render(
      <Provider store={store}>
        <Flyout />
      </Provider>,
    );

    expect(container).toMatchSnapshot();
  });
});
