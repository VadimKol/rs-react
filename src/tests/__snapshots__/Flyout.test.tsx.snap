// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Flyout component > renders correctly 1`] = `
<div>
  <div
    class="_flyout_261b38"
  >
    <h3
      class="_title_261b38"
    >
      2 items are selected
    </h3>
    <div
      class="_buttons_261b38"
    >
      <button
        aria-label="Button"
        class="_button_de8d5c _button_261b38"
        id=""
        type="button"
      >
        Unselect all
      </button>
      <button
        aria-label="Button"
        class="_button_de8d5c _button_261b38"
        id=""
        type="button"
      >
        <a
          class="_download_link_261b38"
          download="2_characters.csv"
        >
          Download
        </a>
      </button>
    </div>
  </div>
</div>
`;
